### PR TITLE
ofi_hmem: new rdma flush function + prov/efa: Emulated Atomic Support for Cuda Devices

### PR DESCRIPTION
**Pull request by a-szegel**
Small change to ofi_hmem.h/cuda_hmem.c to add rdma flush function

Lots of changes to prov/efa to add emulated atomic support for CUDA devices.
1. Added result + fetch memory descriptors to rxr_atomic_ex
2. Switched every memcpy to be ofi_hmem memcpy
3. On the target side, added ofi_hmem memcpy's around emulated atomic impl functions, and performed them on host memory.

Testing:
```
fi_rdm_atomic -D cuda -p efa -E  # passes, and exits cleanly
fi_rdm_atomic -p efa -E                # passes, and exits cleanly
```

### CODE REVIEW

File: src/hmem_cuda.c [hunk 1]
```diff
@@ -760,6 +760,12 @@ bool cuda_is_gdrcopy_enabled(void)
 	return hmem_cuda_use_gdrcopy;
 }
 
+int cuda_flush_rdma_operations(void) {
+	enum cudaFlushGPUDirectRDMAWritesTarget target = cudaFlushGPUDirectRDMAWritesTargetCurrentDevice;
+	enum cudaFlushGPUDirectRDMAWritesScope scope = cudaFlushGPUDirectRDMAWritesToOwner;
+	return cudaDeviceFlushGPUDirectRDMAWrites(target, scope);
+}
```
Comment (shefty) [inline src/hmem_cuda.c:767]: Please fix formatting.  Opening brace for function should be on line by itself.  Limit line lengths to about 80 characters.  target and scope variables don't even look needed.

File: include/ofi_hmem.h [hunk 2]
```diff
@@ -291,6 +291,11 @@ ssize_t ofi_copy_to_hmem_iov(enum fi_hmem_iface hmem_iface, uint64_t device,
 			     size_t hmem_iov_count, uint64_t hmem_iov_offset,
 			     const void *src, size_t size);
 
+int ofi_copy_to_hmem(enum fi_hmem_iface iface, uint64_t device,
+				   void *dest, const void *src, size_t size);
+int ofi_copy_from_hmem(enum fi_hmem_iface iface, uint64_t device,
+				     void *dest, const void *src, size_t size);
```
Comment (shefty) [inline include/ofi_hmem.h:297]: Move this set of changes into a separate patch.  Just move the entire static inline here

**Review by shefty — CHANGES_REQUESTED** [review]
Last commit is also missing signed off by line

File: src/hmem_cuda.c [hunk 3]
```diff
@@ -760,6 +760,12 @@ bool cuda_is_gdrcopy_enabled(void)
...
+
 #else
```
Comment (shijin-aws) [inline src/hmem_cuda.c:770]: Since you declare this new function for both HAVE_CUDA and not HAVE_CUDA, please add a stub function (return `-FI_ENOSYS`) for  `cuda_flush_rdma_operations` in the `#else` block as well .

**Review by shijin-aws — CHANGES_REQUESTED** [review]

File: src/hmem_cuda.c (hunk 1 above)
Comment (a-szegel) [reply src/hmem_cuda.c:767]: fixed

File: include/ofi_hmem.h (hunk 2 above)
Comment (a-szegel) [reply include/ofi_hmem.h]: fixed (gave you 1 more than you asked for)

File: src/hmem_cuda.c (hunk 3 above)
Comment (a-szegel) [reply src/hmem_cuda.c:770]: fixed.

**Comment by a-szegel** [comment]
splitting into two pull requests.
